gio-subclass/src/application.rs

```diff
@@ -46,80 +43,75 @@ impl<S: ApplicationImpl<T>, T: ObjectType + glib::IsA<gio::Application>> Applica
         }
     }
 
     fn parent_open(&self, files: &Vec<gio::File>, hint: &String) {
         unsafe {
             let klass = self.get_class();
             let parent_klass = (*klass).get_parent_class() as *const gio_ffi::GApplicationClass;
-            (*parent_klass).open
-                           .map(|f| {
-                               let n_files = files.len() as i32;
-                                f(self.to_glib_none().0,
-                                  files.to_glib_none().0,
-                                  n_files,
-                                  hint.to_glib_none().0)
-                            })
-                            .unwrap_or(())
+            (*parent_klass)
+                .open
+                .map(|f| {
+                    let n_files = files.len() as i32;
+                    f(
+                        self.to_glib_none().0,
+                        files.to_glib_none().0,
+                        n_files,
+                        hint.to_glib_none().0,
+                    )
+                })
+                .unwrap_or(())
         }
     }
 
-    fn parent_command_line(&self, cmd_line: &gio::ApplicationCommandLine) -> i32{
+    fn parent_command_line(&self, cmd_line: &gio::ApplicationCommandLine) -> i32 {
         unsafe {
             let klass = self.get_class();
             let parent_klass = (*klass).get_parent_class() as *const gio_ffi::GApplicationClass;
-            (*parent_klass).command_line
-                           .map(|f| {
-                               f(self.to_glib_none().0,
-                                cmd_line.to_glib_none().0)
-                            })
-                            .unwrap_or(0)
+            (*parent_klass)
+                .command_line
+                .map(|f| f(self.to_glib_none().0, cmd_line.to_glib_none().0))
+                .unwrap_or(0)
         }
     }
 
-    fn parent_local_command_line(&self, arguments: &Vec<String>) -> Result<i32, ()>{
+    fn parent_local_command_line(&self, arguments: &Vec<String>) -> Option<i32> {
         unsafe {
             let klass = self.get_class();
             let parent_klass = (*klass).get_parent_class() as *const gio_ffi::GApplicationClass;
             let mut exit_status = 0;
-            let success = (*parent_klass).local_command_line
-                           .map(|f| {
-                               f(self.to_glib_none().0,
-                                 arguments.to_glib_none().0, //TODO: conversion
-                                 &mut exit_status)
-                            })
-                            .unwrap_or(0);
-            if success > 0{
-                return Ok(exit_status);
-            }
-            Err(())
+            (*parent_klass)
+                .local_command_line
+                .map(|f| {
+                    f(
+                        self.to_glib_none().0,
+                        arguments.to_glib_none().0, //TODO: conversion
```

Comment:
Which conversion? To `OsString` or ..?

Reply:
Arguments is a  ```*mut *mut *mut c_char```, but there's no length parameter

Reply:
That's fine, it's NULL terminated and the conversion traits ensure that/use that

Reply:
can you paste the error?

Reply:
```
   Compiling gio-subclass v0.1.0 (file:///home/dvhaeren/devel/gtk-subclass/gio-subclass)
error[E0277]: the trait bound `std::string::String: glib::translate::ToGlibContainerFromSlice<'_, *mut *mut *mut i8>` is not satisfied
   --> gio-subclass/src/application.rs:133:35
    |
133 |                         arguments.to_glib_none().0, //TODO: conversion
    |                                   ^^^^^^^^^^^^ the trait `glib::translate::ToGlibContainerFromSlice<'_, *mut *mut *mut i8>` is not implemented for `std::string::String`
    |
    = help: the following implementations were found:
              <std::string::String as glib::translate::ToGlibContainerFromSlice<'a, *mut *const i8>>
              <std::string::String as glib::translate::ToGlibContainerFromSlice<'a, *mut *mut i8>>
    = note: required because of the requirements on the impl of `glib::translate::ToGlibPtr<'_, *mut *mut *mut i8>` for `[std::string::String]`

error[E0277]: the trait bound `std::vec::Vec<std::string::String>: glib::translate::FromGlibPtrBorrow<*mut *mut *mut i8>` is not satisfied
   --> gio-subclass/src/application.rs:409:42
    |
409 |     match imp.local_command_line(&wrap, &from_glib_borrow(arguments)) {
    |                                          ^^^^^^^^^^^^^^^^ the trait `glib::translate::FromGlibPtrBorrow<*mut *mut *mut i8>` is not implemented for `std::vec::Vec<std::string::String>`
    |
    = note: required by `glib::translate::from_glib_borrow`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
error: Could not compile `gio-subclass`.

To learn more, run the command again with --verbose.

```

Reply:
Oh that's because it's not just a `char **` but a `char ***`. It allows changing the array. You need to dereference it at least, and make sure to update the pointer after calling parent class/subclass implementations of this function

Reply:
Ah yes, I misread the glib docs: arguments is indeed ```(inout)```. I'm not sure what that means in terms of ownership. Should the subclass functions return a new ```Vec<String>```. Or do I pass them as a ```&mut```?

Reply:
I think owned makes most sense. The subclass returns a new vector, or just the same vector unchanged. And whatever array is currently behind the pointer is owned by the current code and it would set a new one and free the old one (or leave the old one in place).

Reply:
Or we could indeed just make it `&mut Vec<_>` instead. Code could then still replace the args if it wants to, but most code probably wouldn't care. Probably the more ergonomic approach here

Reply:
But you still need to be sure at the FFI boundaries to update pointers as needed

Reply:
I'd opt for ```&mut Vec<_>```. That seems the most Rust'y to me. Now battling with ffi to figure out how to update those pointers

Reply:
`ptr::read` and `ptr::write` is probably what you want here

Reply:
See commit f2dd812. I'm not sure that is correct...